drivers/clock_control/clock_control_ll_nrfxlib.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <soc.h>
+#include <errno.h>
+#include <device.h>
+#include <kernel_includes.h>
+#include <clock_control.h>
+#include <ble_controller_soc.h>
+#include <multithreading_lock.h>
+
+static int hf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+
+	s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_request(NULL);
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	bool blocking = POINTER_TO_UINT(sub_system);
+
+	if (blocking) {
+		bool is_running = false;
+
+		while (!is_running) {
+			errcode = MULTITHREADING_LOCK_ACQUIRE();
+			if (errcode == 0) {
+				errcode = ble_controller_hf_clock_is_running(
+					&is_running);
+				MULTITHREADING_LOCK_RELEASE();
+			}
+			if (errcode != 0) {
+				return -EFAULT;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int hf_clock_stop(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	s32_t errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_release();
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int hf_clock_get_rate(struct device *dev,
+			     clock_control_subsys_t sub_system, u32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = MHZ(16);
+	return 0;
+}
+
+static int lf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	/* No-op. LFCLFK is started by default. */
+
+	return 0;
+}
+
+static int lf_clock_get_rate(struct device *dev,
+			     clock_control_subsys_t sub_system, u32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = 32768;
+	return 0;
+}
+
+static int clock_control_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* No-op. Initialized by hci_driver_init() in
+	 * subsys/bluetooth/controller/hci_driver.c
+	 */
+
+	return 0;
```

Comment:
What if somebody would like to manage clock before starting Bluetooth?
Please note that system clock driver (which is initialized before kernel and high-level services) requests LFCLK.

Reply:
The SD version of the LFCLK driver should handle an already running LFCLK peripheral AFAIK.

Reply:
But there is only one driver. Either BLE one or System one. So the system will call to this driver before BLE stack will be initialized.

Reply:
Ok, so could we solve this be moving the initialization of the BLE stack to an earlier level?

Reply:
This could be a solution, but you should be aware that many subsystems (like locking, memory management etc.) will might be not operational at this level, so integration could be a bit tricky.

Also, such move may hit us from other side as we could increase boot-up/wake-up time (especially if BLE will wait for clock calibration or stabilization) to unacceptable levels. And there will be no workaround since no other could could run simultaneously before we start multithreading.

I would really like to see BLE stack as consumer, not provider of shared hardware access. I understand that this is something a bit more complex to handle from BLE perspective, but such approach will result is loose bonds between OS and BLE and make all things easier to integrate, maintain and use.

Reply:
Ok, so no locking in earlier initialization. discussed offline and it seems that the best way is to split the initialization, having ble_controller_init in one init level, and the ble config and enable in another.
This way the hardware resources can be accessed earlier.
We can improve upon this for the next release.

Reply:
The initialization is now split, s.t., clock is initialized by `ble_controller_init()` at `PRE_KERNEL_1` time.